Matrix visualization for increase/decrease in percentage for comparison cluster

Anisha1234 opened the pull request:
- This PR includes visualization for describing the percent differences across generated clusters for a run comparison. 
- This will enhance the summary bar graph displayed showing mean values for each cluster.

![main7](https://user-images.githubusercontent.com/16955978/56444593-00c7c300-6317-11e9-823f-72b9a3649e46.gif)

portante commented:
@Anisha1234, we have removed the "`[WIP]`" text from the title of this PR since we have a "`WIP`" label that we use now instead.

Can you rebase  this PR and squash the commits down to one as well?
 
We have moved this PR to the 0.60 release since we are working to push 0.59 out the door shortly.

Anisha1234 commented:
> @Anisha1234, we have removed the "`[WIP]`" text from the title of this PR since we have a "`WIP`" label that we use now instead.
> 
> Can you rebase this PR and squash the commits down to one as well?
> 
> We have moved this PR to the 0.60 release since we are working to push 0.59 out the door shortly.

@portante Ihave rebased and squashed the commits down to one. Moreover the work is ready to be reviewed so removed "WIP" label.

gurbirkalsi commented:
@Anisha1234 I'm seeing NaN values being rendered as part of the percent differences, can you take a look at this error? 
![Pbench_Dashboard](https://user-images.githubusercontent.com/4149652/57320896-0ae71100-70ce-11e9-856e-0f4cb27af135.png)

gurbirkalsi commented:
Can we consider making the percent differences table toggle-able in order to keep the initial rendering of the run comparison page focused on the clustering graphs and matching configurations?

gurbirkalsi commented on web-server/v0.4/src/pages/RunComparison/index.js:
```diff
@@ -295,9 +297,73 @@ class RunComparison extends React.Component {
     } = this.state;
     const { selectedControllers, selectedResults, iterationParams } = this.props;
 
+    const children = [];
```
This logic for generating the column data should be moved to `parse.js` similar to how we invoke `onGenerateIterationClusters` in the event of a cluster change on line 179.

Anisha1234 commented:
@gurbirkalsi I have fixed the suggested changes. Please have a look at it and let me know if any further changes are required.

gurbirkalsi commented on web-server/v0.4/src/utils/parse.js:
```diff
@@ -299,3 +306,65 @@ export const generateIterationClusters = (config, iterations) => {
   return parseClusteredIterations(primaryMetricIterations, clusterLabels, selectedConfig);
 };
+
+export const getComparisionColumn = (record, iterationNo) => {
```
It seems we have a spelling error here - did you mean `getComparisonColumn`?

gurbirkalsi commented on web-server/v0.4/src/utils/parse.js:
```diff
@@ -224,6 +227,10 @@ export const parseClusteredIterations = (clusteredIterations, clusterLabels, sel
             })
           ];
+        const percentage =
+          ((clusterObject[iteration] - clusterObject[0]) / clusterObject[iteration]) * 100;
```
The calculation of percent difference is missing one aspect of the formula where the average of both numbers is used as the divisor. The formula for calculating percent difference is as follows:
`|x1 - x2|/((x1 + x2)/2) * 100`

gurbirkalsi commented:
@Anisha1234 the percent differences don't seem to be rendered here - are you able to reproduce this on your environment? ![Pbench_Dashboard](https://user-images.githubusercontent.com/4149652/58283177-ab4b5f80-7d75-11e9-9b11-4e5bf1ee051c.png)

Anisha1234 commented:
> @Anisha1234 the percent differences don't seem to be rendered here - are you able to reproduce this on your environment? ![Pbench_Dashboard](https://user-images.githubusercontent.com/4149652/58283177-ab4b5f80-7d75-11e9-9b11-4e5bf1ee051c.png)

@gurbirkalsi  I have fixed the issue. It was due to the calculation of no if iterations.

gurbirkalsi commented on web-server/v0.4/src/utils/parse.js:
```diff
@@ -300,3 +316,74 @@ export const generateIterationClusters = (config, iterations) => {
+
+export const getComparisonColumn = (record, iterationNo) => {
+  const children = [];
+  for (let i = 0; i < iterationNo; i += 1) {
```
Can we consider using `forEach()` here to maintain the ES6 usage throughout the codebase? For more on `forEach()`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

Anisha1234 replied:
@gurbirkalsi , the forEach() method executes a provided function for each array element. And here we have to iterate over an integer number. So should we consider using recursion here?

gurbirkalsi replied:
Sorry, I thought `iterationNo` was an array here - this looks good!

gurbirkalsi commented on web-server/v0.4/src/utils/parse.js:
```diff
@@ -300,3 +316,74 @@ export const generateIterationClusters = (config, iterations) => {
+  for (let i = 0; i < iterationNo; i += 1) {
+    children.push({
+      title: `Iteration - ${i}`,
+      dataIndex: '0',
```
In general, we should set unique values here for `dataIndex` and `key` since we are iterating over the length of iterations for a cluster.

gurbirkalsi commented on web-server/v0.4/src/utils/parse.js:
```diff
@@ -300,3 +316,74 @@ export const generateIterationClusters = (config, iterations) => {
+            if (text === 0) {
+              return `${text}%`;
+            }
+            if (text === undefined) {
```
Can this edge case be caught during the parsing phase so we don't need to account for this during the rendering phase?

gurbirkalsi commented on web-server/v0.4/src/utils/parse.js:
```diff
@@ -300,3 +316,74 @@ export const generateIterationClusters = (config, iterations) => {
+              );
+            }
+            if (text === 0) {
+              return `${text}%`;
```
What if we rendered a dash icon here when the percentage is equal to 0 to show this is a baseline representation of the percent difference?

gurbirkalsi commented on web-server/v0.4/src/utils/parse.js:
```diff
@@ -207,6 +209,7 @@ export const parseClusteredIterations = (clusteredIterations, clusterLabels, sel
   const clusteredGraphData = [];
   const graphKeys = [];
   const tableData = [];
+  let noOfIteration = 0;
```
Can we consider naming this variable `maxClusterLength` to better describe this value store as the max number of iterations found out of all clusters?

gurbirkalsi commented:
@Anisha1234 Once we resolve the build error, we should be ready to merge this PR- great work Anisha!

portante commented:
Removing the WIP label as this has been reviewed once already.

portante commented on web-server/v0.4/package.json:
```diff
@@ -118,4 +118,4 @@
     "not ie <= 10"
   ]
-}
+}
```
Why did we remove the trailing newline of this file in this commit?
